Test update

ccoblis opened the pull request.

anastasiamac committed:
First pass at storing cloud images metadata in db and tarball on state server.

anastasiamac committed:
Aligning model with ~/environs/imagemetadata/simplestreams

anastasiamac committed:
Addressed review.

anastasiamac committed:
Full points in comments.

anastasiamac committed:
Renamed version to series. Removed blob and tarball references. Updated model based on one from simple streams.

anastasiamac committed:
Arch is required. Rephrased an err msg.

anastasiamac committed:
Renamed method for clarity.

anastasiamac committed:
Review comments.

anastasiamac committed:
Changed key to series-arch-stream. When stream is empty, "released" is used.

anastasiamac committed:
Corrected test check to go through "naked" find.

anastasiamac committed:
Renamed method for clarity.

anastasiamac committed:
Renamed internal var to "storage".

anastasiamac committed:
Renamed Storage to RootStore.

anastasiamac committed:
Added region, vtype and storage type to key. Changed search to filter using criteria & properties match. Search returns collection.

anastasiamac committed:
Clarity improvements based on review :)

anastasiamac committed:
Adding image id.

anastasiamac committed:
Added comments for clarity.

anastasiamac committed:
Added tests for search clause construction.

anastasiamac committed:
Addressed some review comments - test suite per file, insert/update save, cloud storage properties, func renaming for clarity.

anastasiamac committed:
Using state functions.

anastasiamac committed:
Using state functions.

anastasiamac committed:
Organised imports

anastasiamac committed:
Testing in isolation from state.

anastasiamac committed:
Ensure that glues well with state.

anastasiamac committed:
Behave nicer when metadata cannot be found.

anastasiamac committed:
Don't error when no commit is needed - no changes for update.

waigani committed:
State: separate collections for LastLogin and LastConnection

Separate out LastLogin from the userDoc into it's own collection. Operate directly on docs in this collection without using txns. Do the same for LastConnection in the envUserDoc. No tests where added as existing behaviour should remain the same and is already covered.

Commit:
Fix tests on CentOS

bogdanteleaga commented on cmd/jujud/agent/upgrade_test.go:
```diff
@@ -806,7 +806,18 @@ func (s *UpgradeSuite) assertCommonUpgrades(c *gc.C) {
 	c.Assert(len(args), jc.GreaterThan, 1)
-	c.Assert(args[0], gc.Equals, "apt-get")
+
+	os, err := version.GetOSFromSeries(version.Current.Series)
```
can you use s.packagemanager here?

bogdanteleaga commented on mongo/mongo_test.go:
```diff
@@ -280,12 +280,60 @@ func (s *MongoSuite) TestInstallMongod(c *gc.C) {
 			index = 1
 		}
 		match := fmt.Sprintf(`.* install .*%s`, test.pkg)
+
```
?
is this intended? if not delete it

bogdanteleaga commented on mongo/mongo_test.go:
```diff
@@ -280,12 +280,60 @@ func (s *MongoSuite) TestInstallMongod(c *gc.C) {
+		"centos7", "mongodb*",
+	}
+
+	mockShellCommand(c, &s.CleanupSuite, "yum-config-manager --add-repo")
```
I'm 90% sure you can just mock the command itself without arguments

bogdanteleaga commented on mongo/mongo_test.go:
```diff
@@ -280,12 +280,60 @@ func (s *MongoSuite) TestInstallMongod(c *gc.C) {
+		"37017",
+	}})
+}
+
```
Can you add a test to check the logic on failures?

bogdanteleaga commented on mongo/mongo_test.go:
```diff
@@ -339,7 +397,36 @@ func (s *MongoSuite) TestInstallMongodServiceExists(c *gc.C) {
+			"37017",
+		}})
+	} else {
+		c.Check(cmds, gc.HasLen, 1)
```
This ends up looking really weird. 
Either test this version of cmd's with deepequals, or test that you have above with length(I'd prefer deep equals).

On a second thought, these might be better off as split tests.

EDIT: this seems to actually  test that the service already exists. can you check what the output of cmds should have been on ubuntu?

bogdanteleaga commented on mongo/mongo_test.go:
```diff
@@ -512,6 +599,42 @@ func (s *MongoSuite) TestAddPPAInQuantal(c *gc.C) {
 	}})
 }
 
+func (s *MongoSuite) TestAddPPAInCentOS(c *gc.C) {
```
nitpick: they're not called PPA's on centos, you can call it EPEL or something

bogdanteleaga commented on provider/local/prereqs_test.go:
```diff
@@ -31,6 +31,8 @@ const lsbrelease = `#!/bin/sh
 echo $JUJUTEST_LSB_RELEASE_ID
 `
 
+var package_query string
```
_'s are almost never used in golang

bogdanteleaga commented on provider/local/prereqs_test.go:
```diff
@@ -43,6 +45,20 @@ func init() {
+		}
+	} else {
+		package_query = "dpkg-query"
+	}
```
Maybe do this in SetUp test, and assert that the error is nil

bogdanteleaga commented on provider/local/prereqs_test.go:
```diff
@@ -101,13 +117,14 @@ func (s *prereqsSuite) TestLxcPrereq(c *gc.C) {
 const jujuLocalInstalled = `#!/bin/sh
-if [ "$2" = "juju-local" ]; then return 0; else return 1; fi
+
+if [ "$2" = "juju-local" ] || [ "$5" = "juju-local" ]; then exit 0; else exit 1; fi
```
Add a comment explaining why $2 and $5

bogdanteleaga commented on testing/base.go:
```diff
@@ -217,3 +218,23 @@ func DumpTestLogsAfter(timeout time.Duration, c *gc.C, cleaner TestCleanup) {
+}
+
+func (s *PackageManagerSuite) SetUpTest(c *gc.C) {
+	os, _ := version.GetOSFromSeries(version.Current.Series)
```
Never swallow errors in tests. Assert that it's nil.

bogdanteleaga commented on cmd/jujud/agent/upgrade_test.go:
```diff
@@ -806,7 +806,18 @@ func (s *UpgradeSuite) assertCommonUpgrades(c *gc.C) {
-	c.Assert(args[0], gc.Equals, "apt-get")
+
+	os, err := version.GetOSFromSeries(version.Current.Series)
+	c.Assert(err, gc.IsNil)
```
There's a relatively new(few months old I think) function in juju testing. 

You would do: c.Assert(err, jc.ErrorIsNil).

If jc(juju checkers) is not imported it's usually:
jc "github.com/juju/testing/checkers"

bogdanteleaga commented on version/current_test.go:
```diff
@@ -31,13 +32,23 @@ func (*CurrentSuite) TestCurrentSeries(c *gc.C) {
 		c.Assert(err, gc.IsNil)
 		// There is no lsb_release command on CentOS.
-		switch os {
+		switch current_os {
```
this test doesn't make sense to me. we search for centos in both branches. lsb_release will either error out or it won't, it can't be both.

I would pull the call to getosfromseries outside of the if, and just switch on version.Windows, version.CentOS, etc instead of runtime.GOOS. You need to test this on all 3 platforms though.

bogdanteleaga commented on worker/provisioner/container_initialisation_test.go:
```diff
@@ -289,9 +294,32 @@ func (s *ContainerSetupSuite) assertContainerInitialised(c *gc.C, ctype instance
+	var series string
+	var expected_initial []string
+
+	current_os, _ := version.GetOSFromSeries(version.Current.Series)
```
again, don't swallow errors and don't use _'s

bogdanteleaga commented on worker/provisioner/container_initialisation_test.go:
```diff
@@ -319,30 +347,49 @@ func (s *ContainerSetupSuite) assertContainerInitialised(c *gc.C, ctype instance
+				[]string{"--target-release", "precise-updates/cloud-tools", "cloud-image-utils"}}},
+			{instance.KVM, [][]string{
+				[]string{"uvtool-libvirt"},
+				[]string{"uvtool"}}},
```
I think default here and somewhere else should somehow signal that we're not supposed to run these tests on those machines. AFAIK these were skipped on windows. Also could you refactor these so you don't switch in both places?(you could pass cont as an argument for example).

You could also take a look at the juju/utils/packaging/commander package I think and you could try to extract the proper command from that package based on the series, without relying on strings.

bogdanteleaga commented on worker/provisioner/container_initialisation_test.go:
```diff
@@ -501,15 +548,30 @@ func (s *AddressableContainerSetupSuite) enableFeatureFlag() {
+	for _, test := range cont {
 		s.enableFeatureFlag()
-		s.assertContainerInitialised(c, test.ctype, test.packages, true)
+		s.assertContainerInitialised(c, test, true)
```
same as above

anastasiamac committed:
Sorting out when "not found" makes sense.

anastasiamac committed:
Test for concrete error message.

anastasiamac committed:
Removed redundant AllMetadata.

anastasiamac committed:
Removed redundant AllMetadata.

anastasiamac committed:
Compare structs is so much less messy :D

anastasiamac committed:
Added created time.
Records are immutable and updates are now delete/inserts.

anastasiamac committed:
Separate tests for same and diff image ids

anastasiamac committed:
Order result by date created.

anastasiamac committed:
Added test to check concurrency.
Changed size to numeric type. Size is now non-discriminating in find.

anastasiamac committed:
Comment. Clarity?

Commit:
state: more robust environment death handling in allEnvWatcher

Changed the signature of removed() on the backing structs so that it
takes the env UUID separately from the *State. The *State is now
allowed to be nil.

This allows removal of entities from the store when an entity's
environment as been removed. When allEnvWatcher.Changed() fails to get
a State for an environment it now checks if the environment is still
alive. If it isn't, the entity is removed from the store.

Commit:
state: replace unnecessary panics in allwatcher code

Commit:
Replace unnecessarily complex jcDeepEqualsCheck with some more direct.

And explained the existence of assertEntitiesEqual.

Commit:
state: explain why error isn't returned by backingOpenedPorts.removed

Commit:
state/multiwatcher: added many missing doc comments

Commit:
state/multiwatcher: add deserialisation support for environments

Also added actions which appears to have been missed previously.

Commit:
state/multiwatcher: fixed typo in doc comment

Commit:
state: allow testWatcher to be used by different Backings

Commit:
state: end-to-end test for allEnvWatcher

Commit:
state: statePool improvements

1. Added some explicit unit tests.
2. closeAll now resets the pool so that re-requests work.
3. Add a mutex to protect access.

Commit:
state: removed fragile timing assumptions in allwatcher tests

testWatcher.All() was assuming that deltas would returned with
ShortWait, if there were delta available. This is fragile and is
especially not true with allEnvWatcherStateBacking where State
connections may have to be established (this can take a while). The
tests were failing on the slower CI hosts.

anastasiamac committed:
Added source. Key is back to being attributes amalgamation. More comments.

anastasiamac committed:
Added source. Key is back to being attributes amalgamation. More comments.

Commit:
state: added State.WatchAllEnvs()

This includes the changes to Close() to clean up the allEnvWatcher
plumbing.

Also added a missing unit test for State.Watch()

anastasiamac committed:
Clean up access interface.

anastasiamac committed:
Removed index

mattyw committed:
dependency/engine: Added moreImportant functionality to the dependency
engine.

Commit:
apiserver: added WatchAllEnvs and AllEnvWatcher endpoint

This provides API access to the allEnvWatcher functionality recently
added to state.

Commit:
api: add AllEnvWatcher to facadeVersions

Commit:
apiserver: expose AllEnvWatcher for server only logins

Or else WatchAllEnvs won't work.

Commit:
api: added support for the WatchAllEnvs API

Commit:
Tags, tags everywhere

Commit:
Test fixes

Commit:
Update List API method

Commit:
Added --constranits spaces=, rebasing onto net-cli

Commit:
Return slices from List api methods

Commit:
Use network.Id instead of string for provider id

dooferlad committed:
Depreicated --network in juju deploy

Commit:
Comment update

Commit:
Fixed the bzr revision id of gwacl in dependencies.tsv

dooferlad committed:
Remove tests that use deploy --networks. Remove support for --networks flag (listed, not wired up, just spits out a warning).

jujubot committed:
Merge pull request #2980 from voidspace/spaces-subnets-api

Subnets api: add, create, list

The API side of subnet add, create and list commands.

(Review request: http://reviews.vapour.ws/r/2360/)

wallyworld committed:
Patch version number in test

dooferlad committed:
Linguistic refinement.

jujubot committed:
Merge pull request #3003 from wallyworld/lp-1484993

Patch version number in TestBootstrapReusesAffinityGroupAndVNet

Fixes: https://bugs.launchpad.net/juju-core/+bug/1484993

Patch version number to a fake released version for TestBootstrapReusesAffinityGroupAndVNet


(Review request: http://reviews.vapour.ws/r/2383/)

fwereade committed:
added generic ValueWorker for use in degenerate-worker manifolds; replaced machineLockWorker implementation

jujubot committed:
Merge pull request #3002 from dooferlad/net-cli-add-deployment-support-for-spaces

Net cli add deployment support for spaces



(Review request: http://reviews.vapour.ws/r/2382/)

jujubot committed:
Merge pull request #2982 from mattyw/uniter-dep-engine-conf

dependency/engine: Added config type and moreImportantError support

@fwereade really need to take a look at this to see if it's what he had in mind

(Review request: http://reviews.vapour.ws/r/2362/)

jujubot committed:
Merge pull request #2997 from dimitern/fix-dependencies.tsv-gwacl

Fixed the bzr revision id of gwacl in dependencies.tsv

I'm getting an error when running `godeps -u dependencies.tsv`
on master:

`
godeps: cannot get information on "/home/dimitern/work/go/src/launchpad.net/gwacl": bzr revision-info has unexpected result "244.1.1 andrew.wilkins@canonical.com-20150810080433-4l0x47qsc16ure7c\n"
`

Even though the above is really an issue with godeps itself
(filed a separate MP for this), the current gwacl revision in
dependencies.tsv is wrong - it should be 245, which is the
parent merge commit of 244.1.1 and trunk at the time of landing.

(Review request: http://reviews.vapour.ws/r/2377/)

jujubot committed:
Merge pull request #2989 from TheMue/maas-retry-release-address

Added retry attempt to ReleaseAddress of MAAS provider

Like in EC2 the releasing of an allocated IP address in MAAS is nor retried multiple times.

(Review request: http://reviews.vapour.ws/r/2369/)

dooferlad committed:
Fixed spaces list to return an error per-result.
Some drive-by fixes.

Commit:
Added CanDeallocateAddresses to addresser server API.

dooferlad committed:
* Added life to BackingSubnet
 * Made network.Space.Error opmitempty when json encoding
 * Tidied up spacesAPI.ListSpaces
 * Added a test for if one subnet in a space fails to be fetched rather than all.
Refactored error injection logic for the stub networks fake space; it now shares the same NextErr logic as everything else.

Commit:
Set correct error if environment does not support allocation.

jujubot committed:
Merge pull request #3017 from ericsnowcurrently/value-worker

Merge "ValueWorker" from the maltese-falcon branch.

See http://reviews.vapour.ws/r/2386/ (https://github.com/juju/juju/pull/3006).

(Review request: http://reviews.vapour.ws/r/2397/)

dooferlad committed:
Fixes based on review

Commit:
Added addresser client API with tests.

jujubot committed:
Merge pull request #2995 from mjs/allenvwatcher-merge

Merge the allenvwatcher feature branch into master



(Review request: http://reviews.vapour.ws/r/2375/)

Commit:
Discovered and fixed a few issues with setting/getting constraints and improved tests coverage

anastasiamac committed:
resolved conflicts.

anastasiamac committed:
Comment clarification.

anastasiamac committed:
time is int64 for precision.

anastasiamac committed:
Change copying order.

anastasiamac committed:
Comments.

anastasiamac committed:
Comments.

Commit:
mongo: EnsureServer now always write out mongodb related config files

We had seen reports of the mongod init system config being in place
but the shared secret file not being there. Although it is not
understood exactly how this happens, ensuring that the shared secret
file, SSL key file and monogdb config file are all in place
irrespective of the init system config is more robust and should fix
the reported problem.

Fixes LP #1468579.

davecheney committed:
mongo: make it clear mongo installation depends on operatingsystem

This is a drive by code cleanup that makes it clear that mongodb
installation depends on the operatingsystem, a value poorly proxied
by version.Binary.Series.

Future branches will continue to remove version.Current from the call
chain.

howbazaar committed:
Return the right error.

anastasiamac committed:
Changed find result to be grouped by source.

jujubot committed:
Merge pull request #3029 from howbazaar/fix-logsender-error

Return the right error.



(Review request: http://reviews.vapour.ws/r/2409/)

davecheney committed:
Merge branch 'master' of https://github.com/juju/juju

jujubot committed:
Merge pull request #3030 from mjs/1468579-mongod-not-starting

mongo: EnsureServer now always write out mongodb related config files

We had seen reports of the mongod init system config being in place but the shared secret file not being there. Although it is not understood exactly how this happens, ensuring that the shared secret file, SSL key file and monogdb config file are all in place irrespective of the init system config is more robust and should fix the reported problem.

Fixes LP #1468579.

(Review request: http://reviews.vapour.ws/r/2410/)

anastasiamac committed:
Check storing in db.

anastasiamac committed:
Formatting.

dooferlad committed:
ListSpaces now lists spaces again.

jujubot committed:
Merge pull request #2572 from anastasiamac/image-metadata-bug1452422

Cloud images

Storing metadata in db and tarball on state server (state layer).

(Review request: http://reviews.vapour.ws/r/1945/)

jujubot committed:
Merge pull request #3012 from dooferlad/net-cli-apiserver-subnets-list

Fixed spaces list to return an error per-result.

Fixed spaces list to return an error per-result.
Some drive-by fixes.

(Review request: http://reviews.vapour.ws/r/2392/)

Commit:
Corrected server API to return no error if allocation is not supported.

fwereade committed:
added blockerWorker; used it to coordinate between log-sender and api-caller; verified live, no unit tests yet

jujubot committed:
Merge pull request #2981 from bogdanteleaga/win10

Win10 client support

Support for detecting win10 series.

(Review request: http://reviews.vapour.ws/r/2361/)

Commit:
Addressed review comments

jujubot committed:
Merge pull request #2991 from bogdanteleaga/windows-force-reboot

Force shutdowns on windows

The shutdown command should have -f as a parameter on windows machines to catch some edge cases where it might not execute.

(Review request: http://reviews.vapour.ws/r/2371/)

jujubot committed:
Merge pull request #3026 from dimitern/space-constraints-syntax-net-cli

Added spaces= to --constraints and fixed setting/getting issues

This is an expansion of #2996, which just introduced support for
specifying a list of positive and negative space names in --constraints
with e.g. "spaces=positive1,^negative2,foo,bar,^baz". Negatives have "^"
as prefix, like tags= and (soon to be gone) networks=.

tl;dr; While testing it I've discovered a number of issues with our
constraints handling in state and fixed them.

I took the opportunity to expand the test coverage and document the
behavior around setting deployment and provisioning constraints and
their environment fallbacks. In brief:

 * machine.SetConstraints() behaved differently than state.AddMachine()
   and its relatives, which clear the container= constraint when
   resolving machine provisioning constraints using service deployment
   constraints and environment fallbacks.
 * setConstraintsOp() was not behaving correctly in a few cases,
   omitting values considered empty and thus leaving stale data in
   place, while making it *impossible* to unset any value of type
   `*[]string` once set (e.g. Tags, Networks, and now Spaces). This was
   fixed by making sure we always replace the constraints doc when we
   set it, and we never store or fetch empty values in state. Values are
   considered empty if they pass the double nil check: `v != nil &&
   isZero(*v)`, i.e. having a zero value for its type (slice pointers
   were especially affected - non-nil pointer to an empty slice).
 * fixed a few places where incorrect assumptions were made about how
   constraints work (i.e. relied on the buggy behavior), so while adding
   more tests I also added some comments describing how the constraints
   resolution works and why.

Tested multiple times with both -race and -cover and without.

(Review request: http://reviews.vapour.ws/r/2406/)

dooferlad committed:
Merge remote-tracking branch 'upstream/master' into net-cli

Conflicts:
	api/client.go
	cmd/juju/commands/main.go
	cmd/juju/commands/status.go
	cmd/juju/status/status_test.go
	state/allwatcher_internal_test.go

Commit:
Add volume support for gce.

dooferlad committed:
Fixed imports

dooferlad committed:
Fixed up errors introduced during merge.

Commit:
Introduced finished worker for workers that don't need to run.

fwereade committed:
worker/util.blockerWorker became worker/gate.gate; added all tests

fwereade committed:
address review

jujubot committed:
Merge pull request #3037 from dooferlad/net-cli

Merge upstream/master



(Review request: http://reviews.vapour.ws/r/2418/)

jujubot committed:
Merge pull request #3032 from fwereade/ude9-log-sender-wait-api-password

coordinate log-sender and api-caller workers

  * add gate.Manifold synchronisation mechanism
  * use it to prevent log-sender from trying to connect before api-caller has rewritten agent config

(Review request: http://reviews.vapour.ws/r/2412/)

bogdanteleaga committed:
Fix slice ordering for test

jujubot committed:
Merge pull request #3040 from bogdanteleaga/fix-windows-bug

Fix slice ordering for test

Fixes https://bugs.launchpad.net/juju-core/+bug/1486675

(Review request: http://reviews.vapour.ws/r/2421/)

Commit:
Fix typos in juju help that users see.

dooferlad committed:
Deprecate networks argument in DeployServiceParams.

Commit:
state: improve allwatcher test failure reporting

- Fixed deepEqual so that errors are also treated as comparison failures.
- Made failure output for readable.

Commit:
state: return StatusData consistently from All(Env)Watcher

When an entity's StatusData was empty it would sometimes be returned
as nil and other times as an empty map. This makes the tests hard to
write and was causing intermittent test failures. It also makes
handling the output of the WatchAll and WatchAllEnvs API calls more
difficult for a caller (they had to handle 2 types).

A lack of status data is now always returned as an empty map.

Commit:
state: exported StatePool

statePool was just being used by the AllEnvWatcher plumbing but will
soon be used with the API server too.

Also expanded testing and added a short-circuit for the state server
environment case.

anastasiamac committed:
Cloud image metadata facade.

Commit:
state: added StatePool.SystemState

Commit:
apiserver: use a pool of State instances

... instead of (often) creating a State for each connection. Creating
a new State is somewhat heavyweight and they are goroutine safe so are
safe to share between connections.

jujubot committed:
Merge pull request #3047 from mjs/allenvwatcher-api

api: added support for the WatchAllEnvs API



(Review request: http://reviews.vapour.ws/r/2429/)

jujubot committed:
Merge pull request #3043 from mjs/allwatcher-statusdata-consistency

state: return StatusData consistently from All(Env)Watcher

When an entity's StatusData was empty it would sometimes be returned as nil and other times as an empty map. This makes the tests hard to write and was causing intermittent test failures. It also makes handling the output of the WatchAll and WatchAllEnvs API calls more difficult for a caller (they had to handle 2 types).

A lack of status data is now always returned as an empty map.

This PR also includes improvements to the custom delta comparison function used in the tests. The output has been made more readable.

(Review request: http://reviews.vapour.ws/r/2425/)

dooferlad committed:
Removed a TODO that had been done.
Removed punctuation from end of error message + fixed tests to match.
Dropped SupportsNetworking check from deploy.go

jujubot committed:
Merge pull request #3042 from dooferlad/net-cli-space-contraints-api-apiserver

Deprecate networks argument in DeployServiceParams.



(Review request: http://reviews.vapour.ws/r/2423/)

dooferlad committed:
Fixed localServerSuite.TestNetworkInterfaces - it relied on map ordering.

jujubot committed:
Merge pull request #3052 from dooferlad/fix-map-ordering

Fixed localServerSuite.TestNetworkInterfaces - it relied on map ordering.



(Review request: http://reviews.vapour.ws/r/2434/)

dooferlad committed:
Bump juju/cmd version to fix Windows tests.

jujubot committed:
Merge pull request #3041 from bac/bug-1486640

Fix typos in juju help that users see.

I filed a Launchpad bug (https://bugs.launchpad.net/juju-core/+bug/1486640) detailing some typos that are exposed to users via the command-line 'juju help' commands and in the official documentation hosted at jujucharms.com.  Since the latter is generated from 'juju help' the typos have to be fixed here.

(Review request: http://reviews.vapour.ws/r/2422/)

jujubot committed:
Merge pull request #3055 from dooferlad/deps-bump-fix-windows

Bump juju/cmd version to fix Windows tests.



(Review request: http://reviews.vapour.ws/r/2437/)

Commit:
Commit WiP to discuss tests in MachineSuite.

Commit:
Failure has been to leave tests too early, hmpf.

Commit:
Addressed comments and added tests.

I addressed a review by anastasiamac.
I also added tests for VolumeSource.

Commit:
Spaces CLI work end-to-end with feature tests; added post-net-cli-mvp
feature flag

Commit:
Added testing for the google layer.

Commit:
LP 1486749: Disallow backups for hosted envs

This patch disallows any backups command for a hosted
environment, as it is currently not supported. It prevents
an API connection to the Backups endpoint if not operating
on the system (state server) environment.

The doc for juju backups was updated to indicate that backups
are not supported for hosted environments, but only if the
JES flag is enabled, to avoid confusion for those who are not
running with this feature yet.

Commit:
Final changes.

jujubot committed:
Merge pull request #3058 from dimitern/space-constraints-syntax-net-cli

Spaces CLI working and feature tested end-to-end

Also added post-net-cli-mvp feature flag for the other subcommands of
juju space.

(Review request: http://reviews.vapour.ws/r/2440/)

Commit:
Subnets CLI work end-to-end with feature tests

jujubot committed:
Merge pull request #3028 from davecheney/mongo-remove-series

mongo: make it clear mongo installation depends on operatingsystem

This is a drive by code cleanup that makes it clear that mongodb
installation depends on the operatingsystem, a value poorly proxied
by version.Binary.Series.

Future branches will continue to remove version.Current from the call
chain.

(Review request: http://reviews.vapour.ws/r/2408/)

Commit:
Bumped amz.v3 revision in dependencies.tsv to latest, fixing TestNetworkInterfaces intermittent issue

jujubot committed:
Merge pull request #3038 from TheMue/addresser-cleaner-startup

Addresser cleaner startup

The Addresser is the first worker using a new way to be started as discussed between William, Dimiter, and me. The test is not inside the machine agent anymore. Instead a worker is always started, checks if it is needed or can work (like here based on asking the environment if it is capable of IP addresses deallocation), and then returns its own instance or an instance of the finished worker. The implementation of that is very thin, it ends immediatelly with `err == nil` and is so not being restarted every three seconds.

More workers, like the Firewaller, will follow.

(Review request: http://reviews.vapour.ws/r/2419/)

Commit:
Fixed incorrectly isolated tests

Commit:
Fixed dummy tests to add AZs

Commit:
Fixed the last few tests

jujubot committed:
Merge pull request #3060 from dimitern/subnets-cli-end-to-end

Subnets CLI work end-to-end with feature tests

Like the previous PR #3058 for Spaces CLI, this one adds a feature test
and enables juju subnet list and add commands to use the API.

(Review request: http://reviews.vapour.ws/r/2441/)

jujubot committed:
Merge pull request #3045 from perrito666/add_gce_storage

WIP: Add volume support for gce.

This implements a basic, a bit rough around the edges storage in GCE.


(Review request: http://reviews.vapour.ws/r/2427/)

Commit:
apiserver: removed now unnecessary State cleanup code

Now that a pool of State instances are being used there is no need for
the logic that closes them when an API connection is finished.

jujubot committed:
Merge pull request #3063 from cherylj/backup_state_server_only

LP 1486749: Disallow backups for hosted envs

This patch disallows any backups command for a hosted
environment, as it is currently not supported. It prevents
an API connection to the Backups endpoint if not operating
on the system (state server) environment.

The doc for juju backups was updated to indicate that backups
are not supported for hosted environments, but only if the
JES flag is enabled, to avoid confusion for those who are not
running with this feature yet.

(Review request: http://reviews.vapour.ws/r/2444/)

Commit:
Subnets list CLI fixed and feature tested; some fixes after the live test on EC2

Commit:
apiserver: removed stateResource

Long unused or never used.

Commit:
Final fixes after EC2 live test

Commit:
apiserver: removed State closing infrastructure

State instances should now never be closed once a an API request
finishes because they are shared between requests by the state pool.

davecheney committed:
Merge branch 'master' of https://github.com/juju/juju

Commit:
state: better name for StatePool.ssSt

jujubot committed:
Merge pull request #3064 from mjs/statepool

Use a shared pool of State instances in the API server

Instead of creating new State instances for each API connection, these are now shared.

This PR makes the changes to export the (pre-existing) StatePool from the state package, adding a few extra features and then using it throughout the API server. A lot of API server infrastructure for managing State instances has been removed as it's no longer necessary.

(Review request: http://reviews.vapour.ws/r/2445/)

Commit:
apiserver: allow connections to Dying & Dead environments

We want to allow things like status and logs for Dying or Dead
environments. Operations such as creating machines aren't allowed for
non-alive environments so this is fairly safe to allow.

dooferlad committed:
Merge remote-tracking branch 'upstream/master' into net-cli-2

jujubot committed:
Merge pull request #3068 from dooferlad/net-cli-merge-master

Net cli merge master



(Review request: http://reviews.vapour.ws/r/2449/)

Commit:
Fixed juju subnet list featuretest

jujubot committed:
Merge pull request #3067 from dimitern/subnets-cli-end-to-end

Fixed juju subnet list (added feature test)

The arguments of ListSubnets needed to change to actually do something
useful. A few fixes needed to make the feature test work for juju subnet
list.

During live testing on EC2 a few more changes were necessary to make
sure both add and list work on AWS (e.g. allowing Subnets without ids to
get all of them).

(Review request: http://reviews.vapour.ws/r/2448/)

anastasiamac committed:
Review issues addressed.

jujubot committed:
Merge pull request #3046 from anastasiamac/structured-metadata-facade

WIP: cloud image metadata client.

This PR contains facade and client that support structured cloud image metadata state storage.

This is a WIP since there are still things that will be added to here - the actual apiserver implementation and facade version registration.


(Review request: http://reviews.vapour.ws/r/2428/)

Commit:
Possible fixes for lp:1457130 and lp:1455627

Commit:
Changes after reviews

jujubot committed:
Merge pull request #3073 from dimitern/net-cli-fix-1457130-and-1455627

Possible fixes for lp:1457130 and lp:1455627

Increased the timeout for this frequently flaky test:
pingerSuite.TestAgentConnectionDelaysShutdownWithPing

Hopefully this will make it more reliable on slower
machines/LXC/Windows.

Also improved the test around instancepoller:
MachineSuite.TestManageEnvironRunsInstancePoller

Both were reproducible on my machine, although not always.
With these I fixes I see no more regressions.

(Review request: http://reviews.vapour.ws/r/2454/)

kat-co committed:
Backing out prior change that caused regressions.

This is the change which is being backed out: http://reviews.vapour.ws/r/1716/diff/#

These are the regressions which are caused by the aforementioned patch:
- https://bugs.launchpad.net/juju-core/+bug/1486074
- https://bugs.launchpad.net/juju-core/+bug/1472632

This is the bug which will once again be broken, but it is much lower priority:
- https://bugs.launchpad.net/juju-core/+bug/1456315

jujubot committed:
Merge pull request #3074 from kat-co/fixes-1486074-ssh-identitiesonly-masteres

Backing out prior change that caused regressions.

This is the change which is being backed out: http://reviews.vapour.ws/r/1716/diff/#

These are the regressions which are caused by the aforementioned patch:
- https://bugs.launchpad.net/juju-core/+bug/1486074
- https://bugs.launchpad.net/juju-core/+bug/1472632

This is the bug which will once again be broken, but it is much lower priority:
- https://bugs.launchpad.net/juju-core/+bug/1456315

(Review request: http://reviews.vapour.ws/r/2455/)

Commit:
state: clean up properly in allwatcher tests

The testWatcher used for the allwatcher tests has always been leaving
a running storeManager goroutine running. This hasn't been a problem
until the recent introduction of the allEnvWatcher. With the
allEnvWatcher, if an event arrived after testWatcher.Stop() was called
new State instances would get created causing our test infrastructure
to notice the leaked mongodb connection and panic.

The storeManager is now stopped before closing down the backing store.

jujubot committed:
Merge pull request #3075 from mjs/allenvwatcher-test-state-closing

state: clean up properly in allwatcher tests

The testWatcher used for the allwatcher tests has always been leaving a running storeManager goroutine running. This hasn't actually been much of a problem until the recent introduction of the allEnvWatcher. With the allEnvWatcher, if an event arrived after testWatcher.Stop() was called new State instances would get created causing our test infrastructure to notice the leaked mongodb connection and panic.

The storeManager is now stopped before closing down the backing store.

(Review request: http://reviews.vapour.ws/r/2456/)

jujubot committed:
Merge pull request #3065 from mjs/apiserver-non-live-envs

apiserver: allow connections to Dying & Dead environments

We want to allow things like status and logs for Dying or Dead environments. Operations such as creating machines aren't allowed for non-alive environments so this is fairly safe to allow.

(Review request: http://reviews.vapour.ws/r/2446/)

anastasiamac committed:
Changed search filter to contain collection of architectures and series.

Commit:
state: all(Env)Watcher.TestStateWatcher reliability fixes

The destruction and removal of a machine could be reported in an
unpredictable number of events if allowed to coalesce. Changed the
tests to look for the deltas for each machine destruction step
separately to avoid intermittent failures.

davecheney committed:
container/{kvm,lxc}: use arch.HostArch in place of version.Current.Arch

This PR is a small part of a set that pull apart the confusing mess that
is version.Current.

This PR replaces several uses of version.Current.Arch, and subsiquent
monkey patching of the value with arch.HostArch. In this instance their
usage is synonumous so there is functional change (apart from breaking
the requirement to import version)

Commit:
Fixed constraints handling and improved tests, partially reverting an earlier change

Commit:
Fixed lp:1487727: non-racy check for newAddresser result

rogpeppe committed:
provider/openstack: implement Schema method

jujubot committed:
Merge pull request #3083 from dimitern/lp-1487727-proper-fix-net-cli

Fixed lp:1487727: non-racy check for newAddresser result

Also includes the single commit from #3082.

RB link: http://reviews.vapour.ws/r/2464/

Commit:
Fix 1484789: add-machine fails with hosted envs

Fixed EnvCommandBase to get bootstrap information
from the API if it doesn't exist in the configstore
(just like the system destroy command).

davecheney committed:
gofmt

davecheney committed:
Merge branch 'master' of https://github.com/juju/juju

jujubot committed:
Merge pull request #3077 from davecheney/unpick

container/{kvm,lxc}: use arch.HostArch in place of version.Current.Arch

This PR is a small part of a set that pull apart the confusing mess that
is version.Current.

This PR replaces several uses of version.Current.Arch, and subsequent
monkey patching of the value with arch.HostArch. In this instance their
usage is synonymous so there is functional change (apart from breaking
the requirement to import version)

(Review request: http://reviews.vapour.ws/r/2458/)

davecheney committed:
Merge branch 'master' of https://github.com/juju/juju

anastasiamac committed:
First cut of apiserver impl.

davecheney committed:
all: remove reference to version.Current.Arch

Remove all remaining references to version.Current.Arch and replace
with arch.HostArch(). For tests, replace the version.Current.Arch
patch with a patch of arch.HostArch.

This PR enbles removal of version.Current.Arch and makes it clearer
which code paths explicitly depend on the architecture of the _host_
machine.

jujubot committed:
Merge pull request #2943 from perrito666/fix_1.24_history_worker_tests

Add tests to statushistorypruner worker

statushistorypruner individual functions where unittested but not the worker.
I added tests plus modified PeriodicWorker to be testable with custom timer implementations.

(Review request: http://reviews.vapour.ws/r/2322/)

jujubot committed:
Merge pull request #3090 from juju/net-cli

Net cli has passed

net-cli has merged master and fixed all the reported regressions:
    http://reports.vapour.ws/releases/3002

(Review request: http://reviews.vapour.ws/r/2471/)

davecheney committed:
Fix test failures

wallyworld committed:
Ensure openstack uses correct default block storage

jujubot committed:
Merge pull request #3091 from wallyworld/port-history-worker-tests

Merge pull request #2943 from perrito666/fix_1.24_history_worker_tests

Add tests to statushistorypruner worker

statushistorypruner individual functions where unittested but not the worker.
I added tests plus modified PeriodicWorker to be testable with custom timer implementations.

(Review request: http://reviews.vapour.ws/r/2322/)

(Review request: http://reviews.vapour.ws/r/2472/)

jujubot committed:
Merge pull request #3089 from davecheney/remove-version-current-arch

all: remove reference to version.Current.Arch

Remove all remaining references to version.Current.Arch and replace
with arch.HostArch(). For tests, replace the version.Current.Arch
patch with a patch of arch.HostArch.

This PR enables removal of version.Current.Arch and makes it clearer
which code paths explicitly depend on the architecture of the _host_
machine.

(Review request: http://reviews.vapour.ws/r/2470/)

jujubot committed:
Merge pull request #3092 from wallyworld/default-openstack-storage

Ensure openstack uses correct default block storage

Fixes: https://bugs.launchpad.net/bugs/1484403

Fix referenced bug plus also ec2 tests which were found to be broken.

(Review request: http://reviews.vapour.ws/r/2473/)

waigani committed:
Merge branch 'master' of https://github.com/juju/juju into last-conn

Conflicts:
	state/upgrades.go
	state/upgrades_test.go
	upgrades/steps125.go
	upgrades/steps125_test.go

Commit:
Fixed lp:1455627: Ping timeout varies based on machine load

Commit:
Changes after reviews

jujubot committed:
Merge pull request #3084 from rogpeppe/033-openstack-schema

provider/openstack: implement Schema method

On similar lines to the same method in the local, ec2 and dummy providers.


(Review request: http://reviews.vapour.ws/r/2465/)

jujubot committed:
Merge pull request #3094 from dimitern/fix-lp-1455627-master

Fixed lp:1455627: Ping timeout varies based on machine load

My original attempt to fix this was just by increasing the ping timeout,
but also logging more details about the attempts (PR #3073). However,
despite it failed:

http://reports.vapour.ws/releases/3002/job/xx-run-unit-tests-lxc-wily-amd64/attempt/1959)

it gave some useful logging around each attempt.

This hopefully better fix takes into account the time it normally takes
to open an API connection and call Ping() on it, over 3 attempts, and
using the maximum timeout as ping timeout. This should make the test
pass reliably even on heavily loaded or underpowered machines, while still
verifying the pinger works as expected.

Tested with:

* -check.vv -check.f TestAgentConnectionDelays (normal case)
* like above, but with -race and -cover (simulating heavy load)
* like the first case, but also inducing heave load externally.
* all of the above while running a hangout, which causes loads
in excess of 17 on my machine.

(Review request: http://reviews.vapour.ws/r/2475/)

rogpeppe committed:
update dependency to use bugfixed schema package

jujubot committed:
Merge pull request #3097 from rogpeppe/034-update-schema-dependency

update dependency to use bugfixed schema package

The fix was to an error message.


(Review request: http://reviews.vapour.ws/r/2477/)

Commit:
Address review comment.

Explicitly return nil in the case where ConfigForName
completes successfully.

Commit:
Improved layer separation for gce provider.

Commit:
ProvisioningInfo includes SubnetsToZones; tests pass, pending live tests

Commit:
Fix lp1488581 - patch arch.HostArch

Added in patching of arch.HostArch in a few
tests which failed on non-amd64 archs.

Commit:
Address review comment

Use arch.AMD64 rather than "amd64"

jujubot committed:
Merge pull request #3102 from cherylj/arch_fixes

Fix lp1488581 - patch arch.HostArch

Added in patching of arch.HostArch in a few
tests which failed on non-amd64 archs.

(Review request: http://reviews.vapour.ws/r/2482/)

anastasiamac committed:
Review comments addressed.

anastasiamac committed:
Simplified add to collection as per review comment.

jujubot committed:
Merge pull request #3076 from mjs/allenvwatcher-test-reliability

state: all(Env)Watcher.TestStateWatcher reliability fixes

The destruction and removal of a machine could be reported in an unpredictable number of events if allowed to coalesce. Changed the tests to look for the deltas for each machine destruction step separately to avoid intermittent failures.

(Review request: http://reviews.vapour.ws/r/2457/)

anastasiamac committed:
Added tests and made size optional.

anastasiamac committed:
Registered facade.
Updated integration test to go through api.
Fixed dealing with nil size.

jujubot committed:
Merge pull request #3100 from perrito666/add_gce_storage

Improved layer separation for gce provider.

The existing storage provider for GCE had a layer piercing having compute imported in the wrong places, this is a step towards correcting that.

(Review request: http://reviews.vapour.ws/r/2480/)

davecheney committed:
environs/bootstrap: remove patching of version.Current.Arch

davecheney committed:
Merge branch 'bootstrap-remove-version-current-arch' of https://github.com/davecheney/juju into davecheney-bootstrap-remove-version-current-arch

davecheney committed:
Remove remaining version.Current.Arch patches

rogpeppe committed:
provider/openstack: add appropriate secret attributes to config

jujubot committed:
Merge pull request #3104 from davecheney/davecheney-bootstrap-remove-version-current-arch

environs/bootstrap: remove patching of version.Current.Arch

Remove remaining instances of patching `version.Current.Arch`

(Review request: http://reviews.vapour.ws/r/2484/)

jujubot committed:
Merge pull request #3110 from rogpeppe/035-openstack-secret-attributes

provider/openstack: add appropriate secret attributes to config

I'm not entirely sure if access-key should be considered secret, but
I'm erring on the safe side to start with.



(Review request: http://reviews.vapour.ws/r/2490/)

anastasiamac committed:
Review comments addressed.

jujubot committed:
Merge pull request #3088 from anastasiamac/metadata-facade-apiserver

WIP: Metadata apiserver

This PR contains:
* change to find metadata to accommodate collection of series and architectures;
* apiserver implementation.

WIP because it is still missing:
* tests for apiserver impl;
* declaration of new facade version;
* unexporting of cloud image metadata storage in state;
* feature test modification to go through api layer.

(Review request: http://reviews.vapour.ws/r/2469/)

Commit:
Changes after reviews

Commit:
Disabled a flaky test until we can fix it properly

jujubot committed:
Merge pull request #3106 from dimitern/spaces-constraints-startinstanceparams

Spaces constraints used for provisioning

A few changes around the provisioner worker and API:

* ProvisioningInfo() now includes SubnetsToZones, only filled in when
  spaces constraints are specified.
* Dropped some, but not all (more in a follow-up) support for --networks
  and networks constraints during provisioning (obsoleted by the spaces
  constraints), and related tests.
* environs.StartInstanceParams has a new SubnetsToZones field, populated
  as described above by the provisioner worker task.

Having SubnetsToZones map in StartInstance() will allow the
providers (if they support spaces deployments) to decide which subnets
to use for starting instances where spaces constraints are given.

The EC2 provider will be the first to use SubnetsToZones in an upcoming
PR.

In addition to making sure unit tests pass, manual live testing was done
on local and ec2 providers:

$ juju bootstrap
$ juju space create foo
$ juju deploy ubuntu --constraints spaces=foo,^bar
-> works as before
$ juju deploy mysql --constraints spaces=unknown
-> causes error status to appear on the provisioned machines
$ juju deploy wordpress
-> works as before

(Review request: http://reviews.vapour.ws/r/2486/)

jujubot committed:
Merge pull request #3087 from cherylj/envcmd_config

Fix 1484789: add-machine fails with hosted envs

Fixed EnvCommandBase to get bootstrap information
from the API if it doesn't exist in the configstore
(just like the system destroy command).

(Review request: http://reviews.vapour.ws/r/2468/)

Commit:
Fix lp1489131 - specify hostArch for test

jujubot committed:
Merge pull request #3113 from cherylj/cmd_bootstrap_arch_fix

Fix lp1489131 - specify hostArch for test



(Review request: http://reviews.vapour.ws/r/2492/)

Commit:
state: renamed collections_test.go

To the singular, as per the implementation source file. Also changed
the suite name to match.

Commit:
state: make Inserts into multi-env collections safer

envStateCollection.Insert now performs the required modification of
inserted documents to isolate them from documents for other
environments.

Commit:
state: status history inserts no longer need to include EnvUUID

envStateCollection handles this now.

jujubot committed:
Merge pull request #3117 from mjs/envstatecollection-insert-munging

envStateCollection insert munging

state: renamed collections_test.go
    
...to the singular, as per the implementation source file. Also changed the suite name to match.

---

state: make Inserts into multi-env collections safer
    
envStateCollection.Insert now performs the required modification of inserted documents to isolate them from documents for other environments.

---

state: status history inserts no longer need to include EnvUUID
    
envStateCollection handles this now.


(Review request: http://reviews.vapour.ws/r/2496/)

bz2 committed:
Increment juju to 1.25-beta1

jujubot committed:
Merge pull request #3126 from bz2/inc-1.25-beta1

Increment juju to 1.25-beta1



(Review request: http://reviews.vapour.ws/r/2505/)

dooferlad committed:
EC2 space constraints take into account subnets and availability zones.

dooferlad committed:
Fixed map ordering problems in TestSubnets

sinzui committed:
Master is 1.26-alpha1.

jujubot committed:
Merge pull request #3132 from juju/inc-1.26-alpha1

Master is 1.26-alpha1.



(Review request: http://reviews.vapour.ws/r/2512/)

waigani committed:
Merge branch 'master' of https://github.com/juju/juju into last-conn

jujubot committed:
Merge pull request #2871 from waigani/last-conn

State: separate collections for LastLogin and LastConnection

Separate out LastLogin from the userDoc into it's own collection. Operate directly on docs in this collection without using txns. Do the same for LastConnection in the envUserDoc. No tests where added as existing behaviour should remain the same and is already covered.

(Review request: http://reviews.vapour.ws/r/2250/)

Commit:
state: big multi-env txn handling cleanup

1. multiEnvRunner no longer modifies the []txn.Op it sees in-place. A
new copy is created instead. This is safer and less surprising. Big
comment from Will about this removed.

This lifts a number of restrictions - for example, previously struct
values that needed multi-env adjustments weren't supported.

2. The code for munging txn Inserts to address multi-env concerns has
been greatly simplified by rounding tripping documents through BSON to
bson.D. This also has the benefit of opening up support for document
types that weren't allowed before.

3. The envuser code and tests seemed to be the only area that was
upset about #1. AddEnvironmentUser has been updated to re-read the new
user from the DB so ensure consistent user field values.

4. Moved/removed some dumb and unnecessary internal tests for envuser.

More to come...

dooferlad committed:
Using set to simplify constraints resolution
Fixed tests broken by AssignPrivateIPAddress change in environ.go
More tests for space and subnet constraints

jujubot committed:
Merge pull request #3130 from dooferlad/net-cli-ec2-space-constraints

Net cli ec2 space constraints

EC2 space constraints take into account subnets and availability zones.
Fixed map ordering problems in TestSubnets

(Review request: http://reviews.vapour.ws/r/2510/)

jujubot committed:
Merge pull request #3137 from mjs/multienv-cleanup

state: big multi-env txn handling cleanup

1. multiEnvRunner no longer modifies the []txn.Op it sees in-place. A new copy is created instead. This is safer and less surprising. Big comment from Will about this removed. This lifts a number of restrictions - for example, previously struct values that needed multi-env adjustments weren't supported.

2. The code for munging txn Inserts to address multi-env concerns has been greatly simplified by rounding tripping documents through BSON to bson.D. This also has the benefit of opening up support for document types that weren't allowed before.

3. The envuser code and tests seemed to be the only area that was upset about point 1. AddEnvironmentUser has been updated to re-read the new user from the DB so ensure consistent user field values.

4. Moved/removed some dumb and unnecessary internal tests for envuser.

More to come...

(Review request: http://reviews.vapour.ws/r/2516/)
